sound/soc/intel/boards/sof_rt5682.c

```diff
@@ -29,9 +29,9 @@
 #define SOF_RT5682_MCLK_EN			BIT(3)
 #define SOF_RT5682_MCLK_24MHZ			BIT(4)
 #define SOF_SPEAKER_AMP_PRESENT		BIT(5)
-#define SOF_RT5682_SSP_AMP(quirk)		((quirk) & GENMASK(8, 6))
-#define SOF_RT5682_SSP_AMP_MASK			(GENMASK(8, 6))
 #define SOF_RT5682_SSP_AMP_SHIFT		6
+#define SOF_RT5682_SSP_AMP(quirk)		(((quirk) << SOF_RT5682_SSP_AMP_SHIFT) & GENMASK(8, 6))
```

Comment:
That looks just wrong. You shift by 6 bits and then mask with bits 8..6? WTH?

Reply:
we are using single bitmask for all the different representation : CODEC, CODEC SSP port identification, CLK SRC, from 6th bit i use for Speaker.
if SSP of AMP is 1, we do 1 << 6 and then mask to ensure end result is 1.

Let me relook again with the math. I did few calculations.

Reply:
@sathyap-chrome I added the SOF_RT5682_SSP_AMP_SHIFT based on the GENMASK(8, 6) for the SSP_AMP(quirk). So what you are doing here is incorrect.

I guess you need to verify what the GEN_MASK for the AMP quirk is then fix that and the shift accordingly.

Reply:
@ranj063  Here are prints from my hatch.
ERR kernel: [    6.962845]  lets Prints all values here:
ERR kernel: [    6.962850]  This is hatch - so SSP AMP should be 1 and ssp_codec should be 0
ERR kernel: [    6.962854] sof_rt5682 sof_rt5682: sof_rt5682_quirk = **0x78**
 [Sathya: 1111000 (bit 3,bit 4 both 1. Bit 5 =1 (spk present) BIT 6 =1 (SSP of SPK)
ERR kernel: [    6.962856]  GENMASK (8,6) = 0x1c0
ERR kernel: [    6.962858]  Earlier way of AMP MASK :
ERR kernel: [    6.962860]  SOF_RT5682_SSP_AMP (1 & GENMASK) = 0x0 
[Sathya : this results in SSP of AMP being returned as 0]
ERR kernel: [    6.962861]  Current way of AMP MASK :
ERR kernel: [    6.962863]  SOF_RT5682_SSP_AMP (1 <<6 & Then GENMASK) = 0x40 
[ Sathya : 1000000 [ Bit 6 = 1 ] Now we right shift and we get SSP as 1]

My understanding of the current implementation:
BIT 0 - 2   : used for identifying the SSP of the CODEC
BIT 3        : check on MCLK
BIT 4        : use 24 Mhz MCLK
BIT 5        : is AMP Present
BIT 6-8     : SSP of the AMP 

So the sof_rt5682_quirk should be " **0x78**"  for CML platform with HS codec on SSP0 and SPK AMP on SSP1.

So when setting BIT, we left shift ( (quirk) << SOF_RT5682_SSP_AMP_SHIFT)and then when reading back we right shift. ( (sof_rt5682_quirk & SOF_RT5682_SSP_AMP_MASK) >>
                        SOF_RT5682_SSP_AMP_SHIFT;)



below code snippet for reference: 
+       printk(KERN_ERR " lets Prints all values here:\n");
+       printk(KERN_ERR " This is hatch - so SSP AMP should be 1 and ssp_codec should be 0\n");
+       dev_err(&pdev->dev, "sof_rt5682_quirk = 0x%lx\n", sof_rt5682_quirk);
+
+       printk(KERN_ERR " GENMASK (8,6) = 0x%lx \n", GENMASK (8,6));
+       printk(KERN_ERR " Earlier way of AMP MASK :\n");
+       printk(KERN_ERR " SOF_RT5682_SSP_AMP (1 & GENMASK) = 0x%lx\n", 1 & GENMASK (8,6));
+       printk(KERN_ERR " Current way of AMP MASK : \n");
+       printk(KERN_ERR " SOF_RT5682_SSP_AMP (1 <<6 & Then GENMASK) = 0x%lx\n", (1<<6) & GENMASK (8,6));

Reply:
ok. i guess i found the fix. let me post the v2. the GENMASK in the 1st place is not needed. That should be right way.

Reply:
This looks good to me. BTW,  it seems the GENMASK is not needed on ``#define SOF_RT5682_SSP_CODEC(quirk)		((quirk) & GENMASK(2, 0))``, too.

Reply:
true. But i think we are good with v2 now. Let @ranj063 and Pierre approve :)